pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,385 @@
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.Token == "" {
+		return errors.New("token is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	osName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	teamID, err := strconv.Atoi(c.TeamID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid TeamID, details = %v", err),
+		}
+	}
+
+	var planID int = 0
+	plans, _, err := client.Plans.List(teamID)
+	for _, plan := range plans {
+		if plan.Name == c.Plan {
+			planID = plan.ID
+			break
+		}
+	}
+
+	sshPrivateKey, sshPublicKey, err := NewKey()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate ssh key: %v", err)
+	}
+
+	key, _, err := client.SSHKey.Create(&cherrygo.CreateSSHKey{
+		Label: machine.Name,
+		Key:   sshPublicKey,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating ssh key on cherry server failed: %v", err)
+	}
+
+	serverCreateRequest := cherrygo.CreateServer{
+		ProjectID:   c.ProjectID,
+		Hostname:    machine.Name,
+		Image:       osName,
+		Region:      c.Location,
+		SSHKeys:     []string{strconv.Itoa(key.ID)},
+		IPAddresses: []string{},
+		PlanID:      strconv.Itoa(planID),
+	}
+
+	server, _, err := client.Server.Create(c.ProjectID, &serverCreateRequest)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to create server = %v", err),
+		}
+	}
+
+	status := "pending"
+	var newSrv cherrygo.Server
+	for status != "active" {
+		newSrv, _, _ = client.Server.List(strconv.Itoa(server.ID))
+		status = newSrv.State
+		time.Sleep(30 * time.Second)
+	}
+
+	time.Sleep(30 * time.Second)
+
+	signer, err := ssh.ParsePrivateKey(sshPrivateKey)
+	sshClientConfig := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.PublicKeys(signer)},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	conn, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", newSrv.IPAddresses[0].Address, "22"), sshClientConfig)
```

Comment:
Ugh, not sure how I feel about this

Reply:
I did not have the choice without cloud-init support. Any ideas ?

Reply:
Cherry servers have added cloud-init support to their API. No more tricky ssh stuff